fabtests/prov/lpp/src/hmem_cuda.c

```diff
@@ -0,0 +1,341 @@
+		} else {
+			wait_rx_cq_params.ep_idx = 0;
+			wait_rx_cq_params.context_val = 0x0;
+			wait_rx_cq_params.flags = FI_TAGGED | FI_RECV;
+			TRACE(ri, util_wait_rx_cq(ri, &wait_rx_cq_params));
+
+			verify_buf_params.mr_idx = i;
+			verify_buf_params.length = buffer_lens[i];
+			verify_buf_params.expected_seed = seed_node_a;
+			TRACE(ri, util_verify_buf(ri, &verify_buf_params));
+		}
+	}
+
+	TRACE(ri, util_teardown(ri, pri));
+
+	return 0;
+}
+
+static void send_wait(struct rank_info *ri, struct rank_info *pri,
+							   size_t buf_size, uint64_t tag)
+{
+	struct wait_tx_cq_params wait_tx_cq_params = { 0 };
+
+	INSIST_FI_EQ(ri,
+			fi_tsend(ri->ep_info[0].fid, ri->mr_info[0].uaddr,
+					 buf_size, NULL, pri->ep_info[0].fi_addr, tag, NULL),
+			0);
+
+	wait_tx_cq_params.ep_idx = 0;
+	wait_tx_cq_params.context_val = 0x0;
+	wait_tx_cq_params.flags = FI_TAGGED | FI_SEND;
+	TRACE(ri, util_wait_tx_cq(ri, &wait_tx_cq_params));
+}
+static void recv_wait_validate(struct rank_info *ri, struct rank_info *pri,
+							   size_t buf_size, uint64_t tag)
+{
+	struct wait_rx_cq_params wait_rx_cq_params = { 0 };
+	struct verify_buf_params verify_buf_params = { 0 };
+
+	INSIST_FI_EQ(ri,
+			fi_trecv(ri->ep_info[0].fid, ri->mr_info[0].uaddr,
+					 buf_size, NULL, FI_ADDR_UNSPEC, tag, 0xffff, NULL),
+			0);
+
+	wait_rx_cq_params.ep_idx = 0;
+	wait_rx_cq_params.context_val = 0x0;
+	wait_rx_cq_params.flags = FI_TAGGED | FI_RECV;
+	TRACE(ri, util_wait_rx_cq(ri, &wait_rx_cq_params));
+
+	verify_buf_params.mr_idx = 0;
+	verify_buf_params.length = 0;
+	verify_buf_params.expected_seed = seed_node_a;
+	TRACE(ri, util_verify_buf(ri, &verify_buf_params));
+}
+
+int run_fi_hmem_cuda_sendrecv_d2d(struct rank_info *ri)
+{
+	struct mr_params mr_params = { 0 };
+	struct ep_params ep_params = { 0 };
+	struct rank_info *pri = NULL;
+	uint64_t tag = 0xbeef, buf_size = 1<<14;
+
+	TRACE(ri, util_init(ri));
+
+	mr_params.skip_reg = true;
+	mr_params.hmem_iface = FI_HMEM_CUDA;
+	if (my_node == NODE_A) {
+		mr_params.access = FI_SEND | FI_RECV;
+		mr_params.seed = seed_node_a;
+	} else {
+		mr_params.access = FI_SEND | FI_RECV;
+		mr_params.seed = seed_node_b;
+	}
+	mr_params.length = buf_size;
+	TRACE(ri, util_create_mr(ri, &mr_params));
+
+	ep_params.idx = 0;
+	TRACE(ri, util_create_ep(ri, &ep_params));
+
+	TRACE(ri, util_sync(ri, &pri));
+
+	if (my_node == NODE_A) {
+		send_wait(ri, pri, buf_size, tag);
+		recv_wait_validate(ri, pri, buf_size, tag);
+	} else {
+		recv_wait_validate(ri, pri, buf_size, tag);
+		send_wait(ri, pri, buf_size, tag);
+	}
+
+	TRACE(ri, util_teardown(ri, pri));
+	return 0;
+}
+
+#else
+inline int hmem_cuda_init(void) { return -ENOSYS; }
+void hmem_cuda_cleanup(void){}
+int hmem_cuda_alloc(void *uaddr, size_t len){return -ENOSYS;}
+void hmem_cuda_free(void *uaddr){}
+int hmem_cuda_memcpy_h2d(void *dst, void *src, size_t len){return -ENOSYS;}
+int hmem_cuda_memcpy_d2h(void *dst, void *src, size_t len){return -ENOSYS;}
```

Comment:
fabtests has common hmem support functionality for cuda, rocr, level zero, neuron already

Reply:
> fabtests has common hmem support functionality for cuda, rocr, level zero, neuron already

yes, but this will require a bit of redesign of the lpp regression test.
If that's ok, can this be merge as is, and we will rework this later?